sw/cheri/tests/spi_tests.hh

```diff
@@ -0,0 +1,303 @@
+// Copyright lowRISC Contributors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include "../../common/defs.h"
+#include "../common/console-utils.hh"
+#include "../common/flash-utils.hh"
+#include "../common/sonata-peripherals.hh"
+#include "../common/uart-utils.hh"
+#include "test_runner.hh"
+#include <cheri.hh>
+#include <ds/xoroshiro.h>
+#include <platform-uart.hh>
+
+using namespace CHERI;
+
+/**
+ * Configures the number of test iterations to perform.
+ * This can be overriden via a compilation flag.
+ */
+#ifndef SPI_TEST_ITERATIONS
+#define SPI_TEST_ITERATIONS (1U)
+#endif
+
+/**
+ * Configures the number of random sectors that will be erased in each
+ * iteration of the SPI flash sector erasing test.
+ * This can be overriden via a compilation flag.
+ */
+#ifndef SPI_TEST_FLASH_SECTORS_ERASED
+#define SPI_TEST_FLASH_SECTORS_ERASED (1U)
+#endif
+
+/**
+ * Configures the number of random pages that will be written with
+ * random data in each iteration of the SPI flash random data test.
+ * This can be overriden via a compilation flag.
+ */
+#ifndef SPI_TEST_FLASH_PAGES_WRITTEN
+#define SPI_TEST_FLASH_PAGES_WRITTEN (10U)
+#endif
+
+// The expected JEDEC ID to read from the SPI Flash
+static constexpr uint8_t ExpectedSpiFlashJedecId[3] = {0xEF, 0x40, 0x19};
+
+/**
+ * SPI Flash size definitions. All in bytes.
+ * References to `sector` refer to the smallest erasable sector.
+ *
+ * Sourced from the datasheet:
+ * https://www.winbond.com/resource-files/w25q256jv%20spi%20revg%2008032017.pdf
+ */
+static constexpr unsigned SpiFlashPageSize   = 256;               // 256 B
+static constexpr unsigned SpiFlashSectorSize = 4 * 1024;          // 4 KiB
+static constexpr unsigned SpiFlashSize       = 32 * 1024 * 1024;  // 32 MiB
+static constexpr size_t SpiFlashPages        = SpiFlashSize / SpiFlashPageSize;
+static constexpr size_t SpiFlashSectors      = SpiFlashSize / SpiFlashSectorSize;
+
+/**
+ * Get the start address of a random page in flash.
+ */
+__always_inline uint32_t random_flash_page_addr(ds::xoroshiro::P32R8 &prng) {
```

Comment:
```suggestion
[[always_inline]] uint32_t random_flash_page_addr(ds::xoroshiro::P32R8 &prng) {
```

Also why `always_inline`? Is it important that it's inlined?

Reply:
No - I think I just included this out of habit because it's a small piece of logic that I extracted out of two functions to reduce repetition. I'm happy to get rid of this.